### ISSUE TITLE
Channel feerates

### ISSUE THREAD

**Issue by ZmnSCPxj** [ksedgwic/clboss#3]
CLBOSS has prepared internal hooks for how we can manipulate channel fees, but we need to implement modules that actually change the channel fees.

Currently the CLBOSS just sets the feerates to the weighted median of all the peers of the peer.

* We use median so that extreme settings do not affect us, as the peer might manipulate us by making up artificial peers and creating some channels with those.
* We use a *weighted* median, so that the signal we get is weighted by the size of the channels the peers of the peer have.

Nevertheless, @darosior claims the below on IRC:

```
<zmnscpxj__> the"median of peers of peer" is a reasonable starting point IMO
<darosior> FWIW i don't think it is :)
<zmnscpxj__> how so
<zmnscpxj__> ?
<darosior> From my own experience setting really high fees when you are a consequent router is the more reasonnable
<darosior> Like way more than all your peers
<zmnscpxj__> "consequent router" means what?
<darosior> I mean i've been testing different economic strategies on a ~70-100 channels node for months
<zmnscpxj__> okay, so what does "consequent router" mean? larger than your competitors?
<darosior> Ah, i meant like big enough to do some stats..
<zmnscpxj__> hmmm
<zmnscpxj__> you mean you look at channels with high fee/second and bump their fees up? Is that it?
<zmnscpxj__> or...?
<darosior> I mean i've been increasing the base and ppm fees for a while, nothing based on my peers
<darosior> Actually if i did, my (required atm..) rebalancings would cost me way too much  
<darosior> Hence my plugin
<zmnscpxj__> ah
<zmnscpxj__> but it seems to me that this approximates as well what I described: if you have been getting a lot of fees from outgoing forwards on a channel, that channel is also likely to be depleted and your plugin would bump up its fees by a good amount
* ghost43 has quit (Remote host closed the connection)
<darosior> Yeah, and exponentially discount its usage in the other way
<darosior> Which, i expect would reduce my rebalancing cost
```

To me, this suggests that if you are somehow "larger" than the peers of the peer, you should probably bump up your feerates beyond the baseline, and if you are "smaller", then you should lower your feerates.  Probably the "largeness" value can be based on the capacities of all public channels, just sample all the capacities of every peer of the per (including yourself) and see where you are when this is sorted, getting a 0->1.0 position (0 you are at the lowest ned, 1.0 you are at the highest end, of the capacities of the peers of the peer).  How to translate that number to a feerate multiplier, though, I am uncertain.

Basically, the fee-setting module lets other modules reigster a function that is given the node ID of the peer whose channel fees we are setting.  That function then provides a single `double`, which is multiplied to the basis fees that we derived from the aforementioned weighted median.  A function that does not want to manipulate the channel fee can provide 1.0, which is the identity object in the multiplication monoid.

**Comment by darosior** [comment]
> To me, this suggests that if you are somehow "larger" than the peers of the peer, you should probably bump up your feerates beyond the baseline, and if you are "smaller", then you should lower your feerates.

To complete the above quote, i think so as well. Once again just empirical suboptimal tinkerings but for my personal small node, i can't afford to set fees as high as the "bigger" one i manage. Although i don't need it either as my rebalance cost are way lesser.

**Comment by ZmnSCPxj** [comment]
`FeeModderBySize` and `FeeModderByBalance` implemented, closing.
